File: Pharmaceutical_Company_Example/01_Creational/JavaScript/Singleton.js
class RegulatoryApprovalSystem {
  constructor() {
    if (RegulatoryApprovalSystem._instance) {
      return RegulatoryApprovalSystem._instance;
    }

    this.approvals = {};
    RegulatoryApprovalSystem._instance = this;
  }

  // Add approval status for a drug
  addApproval(drugName, status) {
    this.approvals[drugName] = status;
    console.log(`Added approval status for: ${drugName} - Status: ${status}`);
  }

  // Get approval status for a drug
  getStatus(drugName) {
    if (this.approvals[drugName]) {
      console.log(`Approval status for ${drugName} is: ${this.approvals[drugName]}`);
    } else {
      console.log(`No approval status found for ${drugName}`);
    }
  }

  // Show all approval statuses
  showAllApprovals() {
    console.log("Current approval statuses:");
    for (let drug in this.approvals) {
      console.log(`${drug} : ${this.approvals[drug]}`);
    }
  }
}

// Get the single instance of RegulatoryApprovalSystem
const approvalSystem = new RegulatoryApprovalSystem();

// Add approvals
approvalSystem.addApproval("Test Vaccine 01", "Approved");
approvalSystem.addApproval("Test Small Molecule 01", "Under Review");
approvalSystem.addApproval("Test Monoclonal Antibody 01", "Approved");

approvalSystem.getStatus("Test Small Molecule 01");
approvalSystem.getStatus("Test Vaccine 01");

approvalSystem.showAllApprovals();

// Create another instance (it will return the same instance)
const approvalSystem2 = new RegulatoryApprovalSystem();
approvalSystem2.addApproval("Test Small Molecule 02", "Pending Review");

// Two instances are the same
approvalSystem2.showAllApprovals();
approvalSystem.showAllApprovals();

File: Pharmaceutical_Company_Example/01_Creational/Java/Singleton.java
import java.util.HashMap;
import java.util.Map;

public class RegulatoryApprovalSystem {
    // Static variable to hold the one instance
    private static RegulatoryApprovalSystem instance;

    // Internal storage for approvals
    private Map<String, String> approvals;

    // Private constructor to prevent external instantiation
    private RegulatoryApprovalSystem() {
        approvals = new HashMap<>();
    }

    // Public static method to get the single instance
    public static synchronized RegulatoryApprovalSystem getInstance() {
        if (instance == null) {
            instance = new RegulatoryApprovalSystem();
        }
        return instance;
    }

    // Add approval
    public void addApproval(String drugName, String status) {
        approvals.put(drugName, status);
        System.out.println("Added approval status for: " + drugName + " - Status: " + status);
    }

    // Get status
    public void getStatus(String drugName) {
        if (approvals.containsKey(drugName)) {
            System.out.println("Approval status for " + drugName + " is: " + approvals.get(drugName));
        } else {
            System.out.println("No approval status found for " + drugName);
        }
    }

    // Show all approvals
    public void showAllApprovals() {
        System.out.println("Current approval statuses:");
        for (Map.Entry<String, String> entry : approvals.entrySet()) {
            System.out.println(entry.getKey() + " : " + entry.getValue());
        }
    }

    // Main method for demonstration
    public static void main(String[] args) {
        RegulatoryApprovalSystem approvalSystem = RegulatoryApprovalSystem.getInstance();

        // Add approvals
        approvalSystem.addApproval("Test Vaccine 01", "Approved");
        approvalSystem.addApproval("Test Small Molecule 01", "Under Review");
        approvalSystem.addApproval("Test Monoclonal Antibody 01", "Approved");

        // Check status
        approvalSystem.getStatus("Test Small Molecule 01");
        approvalSystem.getStatus("Test Vaccine 01");

        // Show all approvals
        approvalSystem.showAllApprovals();

        // Create another instance
        RegulatoryApprovalSystem approvalSystem2 = RegulatoryApprovalSystem.getInstance();
        approvalSystem2.addApproval("Test Small Molecule 02", "Pending Review");

        // Verify shared instance
        approvalSystem2.showAllApprovals();
        approvalSystem.showAllApprovals();
    }
}
